### PR TITLE
Make torch._guards, shuffle structures around for migration

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #90636

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_guards.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    # typically happens in GuardBuilder.eval.  In these cases, name is
+    # typically produced by Source.name() (not to be confused with
+    # GuardSource)--morally, we could have stored a Source here.
+    #
+    # Occasionally, name is not a valid Python expression; sometimes
+    # it is meaningless.  Example create_fns that are like this include
+    # GRAD_MODE and SYMBOL_MATCH.
+    name: str
+    source: GuardSource
+    create_fn: Callable[["GuardBuilder", "Guard"], None]
```
Comment (ezyang) [inline torch/_guards.py:46]: so what's the plan for dealing with the GuardBuilder layering problem. Mypy will not be happy until you figure something out.

Comment (voznesenskym) [reply torch/_guards.py:46]: We should drop this to be any callable, nothing about this create_fn is tied to GuardBuilder directly anymore. For the sake of layering, maybe just do Callable[Any, None]. 

Or we move the GuardBuilder functions down to this layer, but I dont wanna do that, at least not yet.

Comment (voznesenskym) [reply torch/_guards.py]: Made it a base class

File: torch/_guards.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+import dataclasses
+import enum
+from typing import Callable, List, Optional
+
+
```
Comment (ezyang) [inline torch/_guards.py:5]: I think this file now deserves a comment saying what it does, and why it lives here and not `torch/_dynamo`

File: torch/_guards.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+        raise NotImplementedError()
+
+    def is_nn_module(self) -> bool:
+        return self in (GuardSource.GLOBAL_NN_MODULE, GuardSource.LOCAL_NN_MODULE)
+
+    def is_local(self):
+        return self in (GuardSource.LOCAL, GuardSource.LOCAL_NN_MODULE)
+
+
+class GuardBuilderBase:
```
Comment (ezyang) [inline torch/_guards.py:29]: Need comment about what this is

**Review by ezyang — APPROVED** [review]

File: torch/_dynamo/output_graph.py [hunk 4]
```diff
@@ -25,13 +25,14 @@
...
 from torch import fx
+from torch._guards import Guard
 from torch.fx.experimental.symbolic_shapes import ShapeEnv
 
 from . import config, logging as torchdynamo_logging, variables
 from .bytecode_transformation import create_instruction, Instruction, unique_id
 from .codegen import PyCodegen
 from .exc import BackendCompilerFailed, unimplemented
-from .guards import Guard, GuardBuilder
+from .guards import GuardBuilder
```
Comment (ezyang) [inline torch/_dynamo/output_graph.py:35]: Consider reexporting Guard from guards so you don't have to wobble all these sites?

Comment (voznesenskym) [reply torch/_dynamo/output_graph.py:35]: I dislike that pattern, I prefer to just point to where things are, not chaining them as transitive deps.

File: torch/_guards.py (hunk 2 above)
Comment (voznesenskym) [reply torch/_guards.py:5]: yes

**Comment by huydhn** [comment]
Here is the linter failure FYI https://hud.pytorch.org/pytorch/pytorch/commit/933b6c4eed675d33274d0bc1dfcb9d8446f412d8

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**